rust_src/src/buffers.rs

```diff
@@ -1122,6 +1161,36 @@ pub unsafe fn per_buffer_idx(offset: isize) -> isize {
     (*obj).as_fixnum_or_error() as isize
 }
 
+/// Delete all overlays of BUFFER and reset its overlay lists.
+#[no_mangle]
+pub extern "C" fn delete_all_overlays(mut buffer: LispBufferRef) {
+    // FIXME: Since each drop_overlay will scan BUF_MARKERS to unlink its
+    // markers, we have an unneeded O(N^2) behavior here.
+
+    let delete_overlays = |overlays: *mut Lisp_Overlay| {
+        if overlays.is_null() {
+            return;
+        }
+
+        let mut overlays_iter = LispOverlayRef::from_ptr(list as *mut c_void)
+            .unwrap_or_else(|| panic!("Invalid overlay reference."))
+            .iter();
+
+        let mut overlay = overlays_iter.next();
+        while let Some(mut ov) = overlay {
```

Comment:
Why is this a `while` and not a `for`? I see `while` as "until we are done, not sure how long that is" and `for` as "for each item in this constrained container".

Reply:
it needs to change `next` field. If we change it in a `for loop`, iterator will terminate immediately.
`while let` allows me to change the item without disturbing the iterator.

Reply:
ah. Makes sense.